kucherov_sergeev/graph_generator.cpp

```diff
@@ -68,11 +188,160 @@ class Graph {
 
   EdgeId get_new_edge_id() { return next_free_edge_id_++; }
 
+  Edge::Color determine_edge_color(const VertexId from_vertex_id,
+                                   const VertexId to_vertex_id) const {
+    const auto from_vertex_depth = get_vertex_depth(from_vertex_id);
+    const auto to_vertex_depth = get_vertex_depth(to_vertex_id);
+
+    if (from_vertex_id == to_vertex_id) {
+      return Edge::Color::Green;
+    }
+    if (get_connected_edge_ids(to_vertex_id).size() == 0) {
+      return Edge::Color::Grey;
+    }
+    if (to_vertex_depth - from_vertex_depth == 1 &&
+        !is_vertices_connected(from_vertex_id, to_vertex_id)) {
+      return Edge::Color::Yellow;
+    }
+    if (to_vertex_depth - from_vertex_depth == 2) {
+      return Edge::Color::Red;
+    }
+
+    throw std::runtime_error("Failed to determine color");
+  }
+
+  void set_vertex_depth(const VertexId vertex_id, const Depth depth) {
+    if (vertex_depths_list_.find(vertex_id) != vertex_depths_list_.end()) {
+      const Depth previous_depth = get_vertex_depth(vertex_id);
+      depth_vertices_list_[previous_depth].erase(
+          std::remove(depth_vertices_list_[previous_depth].begin(),
+                      depth_vertices_list_[previous_depth].end(), vertex_id),
+          depth_vertices_list_[previous_depth].end());
+    }
+
+    depth_vertices_list_[depth].push_back(vertex_id);
+    vertex_depths_list_[vertex_id] = depth;
+  }
+
   VertexId next_free_vertex_id_ = 0;
   EdgeId next_free_edge_id_ = 0;
   std::vector<Vertex> vertices_;
   std::vector<Edge> edges_;
   std::unordered_map<VertexId, std::vector<EdgeId>> adjacency_list_;
+  std::unordered_map<VertexId, Depth> vertex_depths_list_;
+  std::unordered_map<Depth, std::vector<VertexId>> depth_vertices_list_;
```

Comment:
Глубина увеличивается от 0 и вверх. Зачем использовать `unordered_map`? Более эффективно будет использовать обычный `vector`
```suggestion
  std::vector<std::vector<VertexId>> depth_vertices_list_;
```